### PR TITLE
Fix some Array-related performance regressions introduced by the new …

### PR DESCRIPTION
**Pull request by swiftix**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [x] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

…indexing model.

This is a manual specialization of index movement functions for a Strideable index that is required for Array performance.
The optimizer is not capable of creating partial specializations yet.

rdar://25946325

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci Please benchmark

**Comment by swiftix** [comment]
@gribozavr Do you mind reviewing these small changes?

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please benchmark

**Comment by swiftix** [comment]
@swift-ci Please test

File: stdlib/public/core/Arrays.swift.gyb [hunk 1]
```diff
@@ -503,6 +503,86 @@ public struct ${Self}<Element>
...
+    // NOTE: this is a manual specialization of index movement for a Strideable
+    // index that is required for Array performance.  The optimizer is not
+    // capable of creating partial specializations yet.
+    // NOTE: Range checks are not performed here, because it is done later by
+    // the subscript function.
+    return end - start
+  }
+
+  public func _failEarlyRangeCheck(_ index: Int, bounds: Range<Int>) {
+  }
```
Comment (gribozavr) [inline stdlib/public/core/Arrays.swift.gyb]: Could you add a comment that says that this method is a no-op on purpose for performance reasons?

Comment (swiftix) [inline stdlib/public/core/Arrays.swift.gyb]: Sure! Will do.

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
@gribozavr OK to merge? The failure on OS X seems to be unrelated to this change as far as I can see.

**Comment by gribozavr** [comment]
@swiftix Should be OK to merge.

**Pull request merged**